connectors/sources/github.py

```diff
@@ -33,57 +120,115 @@ def __init__(self, configuration):
+        # Adding a 5 second delay to account for server delays
+        return (reset - time.time()) + 5
+
+    async def _put_to_sleep(self):
+        retry_after = await self._get_retry_after()
+        logger.debug(f"Connector will attempt to retry after {retry_after} seconds.")
+        await self._sleeps.sleep(retry_after)
+        raise Exception("Rate limit exceeded.")
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def get_response(self, resource):
         """Execute request using getitem method of GitHubAPI.
 
         Args:
             resource (str): API to get the response
 
         Returns:
-            list: Response of the request
+            dict/list: Response of the request
         """
-        return await self._get_client.getitem(url=resource)
+        try:
+            return await self._get_client.getitem(url=resource)
+        except BadRequest as exception:
+            if exception.status_code == 403 and "rate limit" in str(exception).lower():
+                await self._put_to_sleep()
+            elif self.is_unauthorized(exception=exception):
+                raise UnauthorizedException(
+                    "Your Github token is either expired or revoked. Please check again."
+                ) from exception
+            else:
+                raise
+        except Exception:
+            raise
 
-    async def execute_request(self, method, url):
-        """Execute request using _request method of GitHubAPI.
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def get_response_headers(self, method, url):
+        """Execute request using _request method of GitHubAPI and returns response headers.
 
         Args:
-            request (str): API for the request
+            method (str): Method of the request
+            url (str): API to get the response
 
         Returns:
-            list: Response of the request
+            dict: Response of the headers
         """
         status_code, response_headers, data = await self._get_client._request(
-            method="get", url=USER_API, headers=self.headers
+            method=method, url=url, headers=self.headers
         )
         if status_code == 200:
             return response_headers
-        data = data.decode("utf-8")
-        message = json.loads(data)
-        raise Exception(
-            f"Something went wrong while executing request Exception: {message['message']}"
-        )
+        try:
+            data = data.decode("utf-8")
+            message = json.loads(data)
+        except ValueError as exception:
+            logger.exception(
+                f"Error while loading the response data. Error: {exception}"
+            )
+            raise
+        if status_code == 403 and "rate limit" in str(message).lower():
```

Comment:
As noted above, probably useful to use a variable here or even better use a reusable method.